tests/api_fastapi/execution_api/routes/test_task_instance.py

```diff
@@ -0,0 +1,192 @@
+
+    def test_ti_update_state_conflict_if_not_queued(self, client, session, create_task_instance):
+        """
+        Test that a 409 error is returned when the Task Instance is not in a state where it can be marked as
+        running. In this case, the Task Instance is first in NONE state so it cannot be marked as running.
+        """
+        ti = create_task_instance(
+            task_id="test_ti_update_state_conflict_if_not_queued",
+            state=State.NONE,
+        )
+        session.commit()
+        test_client = client(apps="execution")
+
+        response = test_client.patch(
+            f"/execution/task_instance/{ti.id}/state",
+            json={
+                "state": "running",
+                "hostname": "random-hostname",
+                "unixname": "random-unixname",
+                "pid": 100,
+                "start_date": "2024-10-31T12:00:00Z",
+            },
+        )
+
+        assert response.status_code == 409
+        assert response.json() == {
+            "detail": {
+                "message": "TI was not in a state where it could be marked as running",
+                "previous_state": State.NONE,
+                "reason": "invalid_state",
+            }
+        }
+
+        assert session.scalar(select(TaskInstance.state).where(TaskInstance.id == ti.id)) == State.NONE
+
+    @pytest.mark.parametrize(
+        "state, end_date, expected_state",
+        [
+            (State.SUCCESS, "2024-10-31T12:00:00Z", State.SUCCESS),
+            (State.FAILED, "2024-10-31T12:05:00Z", State.FAILED),
+            (State.SKIPPED, "2024-10-31T12:05:00Z", State.SKIPPED),
+            (State.SKIPPED, None, State.SKIPPED),
+            (State.UPSTREAM_FAILED, None, State.UPSTREAM_FAILED),
+            (State.REMOVED, None, State.REMOVED),
+        ],
+    )
+    def test_ti_update_state_to_terminal(
+        self, client, session, create_task_instance, state, end_date, expected_state
+    ):
+        ti = create_task_instance(
+            task_id="test_ti_update_state_to_terminal",
+            state=State.RUNNING,
+        )
+        session.commit()
+        test_client = client(apps="execution")
+
+        payload = {"state": state}
+        if end_date:
+            payload["end_date"] = end_date
+
+        response = test_client.patch(
+            f"/execution/task_instance/{ti.id}/state",
+            json=payload,
+        )
+
+        assert response.status_code == 204
+        assert response.text == ""
+
+        assert session.scalar(select(TaskInstance.state).where(TaskInstance.id == ti.id)) == expected_state
+
+        if expected_state == State.SKIPPED and not end_date:
+            assert (
+                session.scalar(select(TaskInstance.end_date).where(TaskInstance.id == ti.id)) == ti.start_date
+            )
+        elif expected_state in [State.UPSTREAM_FAILED, State.REMOVED]:
+            assert session.scalar(select(TaskInstance.end_date).where(TaskInstance.id == ti.id)) is None
```

Comment:
```suggestion
        ti = session.get(TaskInstance, ti.id)
        assert ti.state == expected_state
        assert ti.end_date.isoformat() == end_date
```

I think? It's a bit hard to grok what this test is asserting with the if/else in the tests, I'd rarther we made it fully table driven.

Reply:
Updated